File: nexus_banking_networks/nexus_network/advanced_banking_system/java/Blockchain.java
// Blockchain.java
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;

public class Blockchain {
    private List<Block> chain;
    private int difficulty;

    public Blockchain(int difficulty) {
        this.chain = new ArrayList<>();
        this.difficulty = difficulty;
        this.chain.add(createGenesisBlock());
    }

    private Block createGenesisBlock() {
        // Create the genesis block of the blockchain
        return new Block("Genesis Block", "0");
    }

    public Block getLatestBlock() {
        // Get the latest block in the blockchain
        return chain.get(chain.size() - 1);
    }

    public void addBlock(String data) {
        // Add a new block to the blockchain
        Block newBlock = new Block(data, getLatestBlock().getHash());
        newBlock.mineBlock(difficulty);
        chain.add(newBlock);
    }
}

class Block {
    private String data;
    private String previousHash;
    private String hash;

    public Block(String data, String previousHash) {
        this.data = data;
        this.previousHash = previousHash;
        this.hash = calculateHash();
    }

    private String calculateHash() {
        // Calculate the hash of the block
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            String data = this.data + this.previousHash;
            byte[] bytes = md.digest(data.getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : bytes) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException(e);
        }
    }

    public void mineBlock(int difficulty) {
        // Mine the block to find a valid hash
        while (!hash.substring(0, difficulty).equals(new String(new char[difficulty]).replace('\0', '0'))) {
            hash = calculateHash();
        }
    }
}

File: blockchain_integration/pi_network/pi-stablecoin/models/blockchain.py
import hashlib
import time

class Block:
  def __init__(self, index, previous_hash, timestamp, data):
    self.index = index
    self.previous_hash = previous_hash
    self.timestamp = timestamp
    self.data = data
    self.hash = self.calculate_hash()

  def calculate_hash(self):
    data_string = str(self.index) + self.previous_hash + str(self.timestamp) + str(self.data)
    return hashlib.sha256(data_string.encode()).hexdigest()

class Blockchain:
  def __init__(self):
    self.chain = [self.create_genesis_block()]

  def create_genesis_block(self):
    return Block(0, "0", int(time.time()), "Genesis Block")

  def get_latest_block(self):
    return self.chain[-1]

  def add_block(self, new_block):
    new_block.previous_hash = self.get_latest_block().hash
    new_block.hash = new_block.calculate_hash()
    self.chain.append(new_block)
